src/crimson/os/seastore/object_data_handler.h

```diff
@@ -23,6 +23,8 @@ struct ObjectDataBlock : crimson::os::seastore::LogicalCachedExtent {
     : LogicalCachedExtent(std::move(ptr)) {}
   ObjectDataBlock(const ObjectDataBlock &other)
     : LogicalCachedExtent(other) {}
+  ObjectDataBlock(extent_len_t length)
+    : LogicalCachedExtent(length) {}
```

Comment:
> which leads to constructions for non-fully-loaded extents are actually constructing extents with arbitrary CachedExtent::ptr

Without this public constructor, the code is still compiling seems due to that `Cache` is a friend of `CachedExtent` at https://github.com/ceph/ceph/blob/46a4fbe60348abd9837471dc78b8e7ee1a5de738/src/crimson/os/seastore/cache.h#L943

And in that case, the same constructor is called: https://github.com/ceph/ceph/blob/46a4fbe60348abd9837471dc78b8e7ee1a5de738/src/crimson/os/seastore/cached_extent.h#L739-L742

So the original code won't cause "arbitrary CachedExtent::ptr" because `std::optional<ceph::bufferptr> ptr;` will be default initialized to `std::nullopt`. Is my understanding correct?

Anyway, I think it's fine to keep the constructor explicit at the `ObjectDataBlock` level.

Reply:
> So the original code won't cause "arbitrary CachedExtent::ptr" because `std::optional<ceph::bufferptr> ptr;` will be default initialized to `std::nullopt`. Is my understanding correct?

I think the reason why, without this constructor, the compiling still succeeds is that `ObjectDataBlock::ObjectDataBlock(ceph::bufferptr)` is resolved, because `ceph::bufferptr` can be constructed with a single `uint32_t` value.

Reply:
That's unexpected, `explicit` keyword can be used to avoid the confusion here.
`explicit ObjectDataBlock(ceph::bufferptr &&ptr)`

Reply:
Done.